File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      is_blank?(input) == "" -> "Fine. Be that way!"
      is_question?(input) -> "Sure."
      is_yelling?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp is_blank?(str) do
    String.strip(str)
  end

  defp is_question?(str) do
    String.ends_with?(str, "?")
  end

  defp is_yelling?(str) do
    str == String.upcase(str) and str != String.downcase(str)
  end
end

File: go/bob/bob.go
// bob the lackadaisical teenager
package bob

import "strings"

// get bob's response to a remark
func Hey(remark string) string {
	r := strings.TrimSpace(remark)

	if isSilence(r) {
		return "Fine. Be that way!"
	}

	if isYelling(r) {
		if isQuestion(r) {
			return "Calm down, I know what I'm doing!"
		} else {
			return "Whoa, chill out!"
		}
	}

	if isQuestion(r) {
		return "Sure."
	}

	return "Whatever."
}

func isSilence(remark string) bool {
	return len(remark) == 0
}

func isYelling(remark string) bool {
	return strings.ToUpper(remark) == remark && strings.ToLower(remark) != remark
}

func isQuestion(remark string) bool {
	return strings.HasSuffix(remark, "?")
}
